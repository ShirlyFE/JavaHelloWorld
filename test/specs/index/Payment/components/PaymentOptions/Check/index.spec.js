import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import { Check } from 'index/Payment/components/PaymentOptions/Check';
import InputNumeric from 'react-base-ui/lib/components/InputNumeric';
import Dropdown from 'react-base-ui/lib/components/Dropdown';
import resetActions from 'utils/resetActions';

jest.mock('index/Payment/components/PaymentOptions/utils/payment', () => ({
  getDefaultAmount: jest.fn(() => (300)),
  getRemaining: jest.fn((index, amount) => amount <= 300 ? -50 : {
    remaining: `$${amount}`
  })
}));

describe('index/Payment/components/PaymentOptions/Check', () => {
  const actions = {
    changeCheckAmount: jest.fn(),
    changeRemaining: jest.fn(),
    updateCheckNumber: jest.fn(),
    clearOptionAndPaymentErrs: jest.fn(),
    onChange: jest.fn()
  };

  const setup = (props) => mount(<Check {...props} />);

  afterEach(() => {
    resetActions(actions);
  });

  it('Check component should render without error', () => {
    const props = {
      item: {
        list: [
          { value: '172', name: 'Credit' },
          { value: '173', name: 'Check' }
        ],
        amount: 0,
        formatCheckAmount: 300,
        errors: [
          {
            key: 0,
            name: 'checkNumber'
          }
        ]
      },
      value: '173',
      index: 0,
      optionLen: 3,
      children: (<span>Check info</span>),
      isRefund: false,
      ...actions
    };
    const component = setup(props);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Check component in refund workflow should render without error', () => {
    const props = {
      item: {
        list: [
          { value: '172', name: 'Credit' },
          { value: '173', name: 'Check' }
        ],
        amount: 200,
        formatCheckAmount: 200,
        checkNumber: '623'
      },
      value: '173',
      index: 0,
      optionLen: 3,
      children: (<span>Check info</span>),
      isRefund: true,
      ...actions
    };

    const component = setup(props);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Check component change payment option and amount should work fine', () => {
    const props = {
      item: {
        list: [
          { value: '172', name: 'Credit' },
          { value: '173', name: 'Check' }
        ],
        amount: 200,
        formatCheckAmount: 300,
        checkNumber: ''
      },
      value: '173',
      index: 0,
      optionLen: 3,
      children: (<span>Check info</span>),
      isRefund: false,
      ...actions
    };

    const component = setup(props);

    expect(component.find('.payment-check')).toHaveLength(1);
    expect(component.find('.payment-check-number')).toHaveLength(1);
    expect(component.find('#checkAmount')).toHaveLength(1);

    const confirmButton = component.find('.payment-amount-exception-alert').find('button.btn-strong');
    confirmButton.simulate('click');

    const paymentOptionDropdown = component.find(Dropdown);
    expect(paymentOptionDropdown).toHaveLength(1)
    paymentOptionDropdown.node.props.onChange({});
    expect(actions.onChange).toBeCalled();

    const checkNumInput = component.find('.payment-check-number').find('input');
    checkNumInput.simulate('blur');
    expect(actions.updateCheckNumber).toHaveBeenCalled();

    const amountInput = component.find(InputNumeric);
    const amountInputHtmlElement = component.find('#checkAmount');
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).not.toHaveBeenCalled();
    expect(actions.changeRemaining).not.toHaveBeenCalled();

    confirmButton.simulate('click');
    expect(actions.changeCheckAmount).toHaveBeenCalledTimes(1);
    expect(actions.changeRemaining).toHaveBeenCalledTimes(1);

    amountInput.node.input.value = 400;
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).toHaveBeenCalledTimes(2);
    expect(actions.changeRemaining).toHaveBeenCalledTimes(2);
  });

  it('Check component refund process should work fine', () => {
    const props = {
      item: {
        list: [
          { value: '172', name: 'Credit' },
          { value: '173', name: 'Check' }
        ],
        amount: 200,
        formatCheckAmount: 300,
      },
      value: '173',
      index: 0,
      optionLen: 3,
      children: (<span>Check info</span>),
      isRefund: true,
      ...actions
    };

    const component = setup(props);

    expect(component.find('.payment-check')).toHaveLength(1);
    expect(component.find('.payment-check-number')).toHaveLength(0);
    expect(component.find('#checkAmount')).toHaveLength(1);

    const amountInput = component.find(InputNumeric);
    const amountInputHtmlElement = component.find('#checkAmount');
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).not.toHaveBeenCalled();
    expect(actions.changeRemaining).not.toHaveBeenCalled();

    resetActions(actions);
    const confirmButton = component.find('.payment-amount-exception-alert').find('button.btn-strong');
    confirmButton.simulate('click');
    expect(actions.changeCheckAmount).toHaveBeenCalled();
    expect(actions.changeRemaining).toHaveBeenCalled();

    resetActions(actions);
    amountInput.node.input.value = 400;
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).toHaveBeenCalled();
    expect(actions.changeRemaining).toHaveBeenCalled();
  });

  it('Check component should work fine even if meets error', () => {
    const props = {
      item: {
        list: [
          { value: '172', name: 'Credit' },
          { value: '173', name: 'Check' }
        ],
        amount: 200,
        formatCheckAmount: 300,
        errors: [
          { message: 'mock check error', name: '' },
          { message: 'mock check number error', name: 'checkNumber' }

        ]
      },
      value: '173',
      index: 0,
      optionLen: 3,
      children: (<span>Check info</span>),
      ...actions
    };

    const component = setup(props);

    expect(component.find('.payment-check')).toHaveLength(1);
    expect(component.find('.payment-check-number')).toHaveLength(1);
    expect(component.find('#checkAmount')).toHaveLength(1);
    expect(component.find('.payment-option-list--error')).toHaveLength(1);

    const amountInput = component.find(InputNumeric);
    const amountInputHtmlElement = component.find('#checkAmount');
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).not.toHaveBeenCalled();
    expect(actions.changeRemaining).not.toHaveBeenCalled();

    resetActions(actions);
    const checkNumberInput = component.find('.payment-check-number').find('input');
    checkNumberInput.simulate('focus');
    expect(actions.clearOptionAndPaymentErrs).toHaveBeenCalledTimes(1);
    expect(actions.updateCheckNumber).not.toHaveBeenCalled();
    checkNumberInput.simulate('blur', { value: '123' }, true);
    expect(actions.clearOptionAndPaymentErrs).toHaveBeenCalledTimes(2);
    expect(actions.updateCheckNumber).toHaveBeenCalled();

    resetActions(actions);
    const confirmButton = component.find('.payment-amount-exception-alert').find('button.btn-strong');
    confirmButton.simulate('click');
    expect(actions.changeCheckAmount).toHaveBeenCalled();
    expect(actions.changeRemaining).toHaveBeenCalled();

    resetActions(actions);
    amountInput.node.input.value = 400;
    amountInputHtmlElement.simulate('blur');
    expect(actions.changeCheckAmount).toHaveBeenCalled();
    expect(actions.changeRemaining).toHaveBeenCalled();
  });
});
