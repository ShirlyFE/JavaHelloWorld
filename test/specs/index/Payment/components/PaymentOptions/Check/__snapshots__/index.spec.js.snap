// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index/Payment/components/PaymentOptions/Check Check component in refund workflow should render without error 1`] = `
<Check
  changeCheckAmount={[Function]}
  changeRemaining={[Function]}
  clearOptionAndPaymentErrs={[Function]}
  index={0}
  isRefund={true}
  item={
    Object {
      "amount": 200,
      "checkNumber": "623",
      "formatCheckAmount": 200,
      "list": Array [
        Object {
          "name": "Credit",
          "value": "172",
        },
        Object {
          "name": "Check",
          "value": "173",
        },
      ],
    }
  }
  onChange={[Function]}
  optionLen={3}
  updateCheckNumber={[Function]}
  value="173"
>
  <div
    className="payment-option payment-check"
  >
    <div
      className="payment-option-list aaui-flexbox"
    >
      <div
        className="aaui-flexbox aaui-flexbox-c1"
      >
        <Dropdown
          autoOpen={false}
          data={
            Array [
              Object {
                "name": "Credit",
                "value": "172",
              },
              Object {
                "name": "Check",
                "value": "173",
              },
            ]
          }
          data-qa-id="payment-paymentOption-optionList"
          filterPlaceholder="Filter..."
          fuzzyQuery={false}
          noResult="No results found."
          onChange={[Function]}
          placeholder="Select one..."
          showAll={true}
          showDeselectall={false}
          showTxtOnlyCheckedOneItem={true}
          theme="flat"
          txtSuffix="selected"
          value="173"
          visible={true}
        >
          <div
            className="aaui-dropdown"
            data-qa-id="payment-paymentOption-optionList"
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
          >
            <button
              className="aaui-dropdown__button aaui-dropdown-btn-flat"
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              role="button"
              type="button"
            >
              <div
                className="aaui-dropdown__button-text"
                title=""
              >
                 Select one...
              </div>
              <i
                className="aaui-dropdown__button-icon icon-chevron-down"
              />
            </button>
            <div
              aria-expanded={false}
              className="aaui-dropdown__menu-wrapper  u-hidden"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="combobox"
              tabIndex={0}
            >
              <div
                className="aaui-dropdown__menu"
              >
                <ul
                  aria-expanded={false}
                  className=""
                  role="listbox"
                >
                  <li
                    className=""
                    key="172"
                    onClick={[Function]}
                    role="option"
                  >
                    <span />
                  </li>
                  <li
                    className=" is-selected"
                    key="173"
                    onClick={[Function]}
                    role="option"
                  >
                    <span />
                  </li>
                </ul>
                <div>
                  <div
                    className="aaui-dropdown__menu-footer"
                  >
                    <div
                      className="u-hidden"
                    >
                      <i
                        className="icon icon-spinner"
                      />
                       Loading...
                    </div>
                    <div />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </Dropdown>
      </div>
      <div
        className="aaui-flexbox aaui-flexbox-c2"
      >
        <label
          className="payment-symbol"
          htmlFor="checkAmount"
        >
          $
        </label>
        <InputNumeric
          allowBlank={true}
          allowKeySpin={true}
          allowMouseSpin={false}
          className=""
          currency="USD"
          data-qa-id="payment-paymentOption-optionAmount"
          decimals={2}
          disabled={false}
          icon=""
          iconHint=""
          id="checkAmount"
          increment={1}
          items={Array []}
          listClass="input-numeric-list"
          listClassName=""
          listConfig={Object {}}
          listMaxHeight="300px"
          listPopupOptions={Object {}}
          listWidth="auto"
          max={9999999999.99}
          min={0}
          name=""
          onBlur={[Function]}
          onFocus={[Function]}
          onListClose={[Function]}
          onListOpen={[Function]}
          onTextChange={[Function]}
          onValueChange={[Function]}
          placeHolder=""
          prefix="an"
          showClear={false}
          showGroup={false}
          showSpinner={false}
          showTrigger={false}
          showTrigger2={false}
          size="md"
          textAlign="right"
          triggerIcon="icon-chevron-down"
          triggerIcon2="icon-chevron-down"
          triggerIconHint=""
          triggerIconHint2=""
          triggerIconToggle="icon-chevron-up"
          triggerIconToggle2="icon-chevron-up"
          value={200}
        >
          <div
            className="an-input-numeric input-group buttons-0 input-group--md"
          >
            <input
              className="input input-group__field textalign-right input__field"
              data-qa-id="payment-paymentOption-optionAmount"
              disabled={false}
              id="checkAmount"
              name=""
              onBlur={[Function]}
              onChange={[Function]}
              onDragEnter={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
              placeholder=""
              type="text"
            />
          </div>
        </InputNumeric>
        <span>
          Check info
        </span>
      </div>
    </div>
    <Alert
      className="payment-amount-exception-alert"
      confirmText="OK"
      onConfirm={[Function]}
      title="System Message"
      type="alert"
    >
      <Modal
        className="modal--alert payment-amount-exception-alert"
        onClose={[Function]}
        shown={false}
        title="System Message"
      >
        <div
          className="modal-wrap"
          role="dialog"
        >
          <div
            className="modal modal--alert payment-amount-exception-alert"
            onKeyDown={[Function]}
          >
            <section
              className="modal-box"
            >
              <header
                className="modal-header"
              >
                <h3
                  className="modal-title"
                >
                  System Message
                </h3>
                <span
                  aria-label="close-dialog"
                  className="icon-close modal-close"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  tabIndex="0"
                />
              </header>
              <div
                className="modal-body"
              />
              <div
                className="modal-footer"
              >
                <Button
                  className=""
                  loading={false}
                  menuData={Array []}
                  noSubmit={false}
                  onClick={[Function]}
                  type="strong"
                >
                  <button
                    className="btn btn-strong"
                    menuData={Array []}
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onMouseOver={[Function]}
                    type="submit"
                  >
                    <span>
                      OK
                    </span>
                  </button>
                </Button>
              </div>
            </section>
          </div>
          <div
            className="modal-mask"
          />
        </div>
      </Modal>
    </Alert>
  </div>
</Check>
`;

exports[`index/Payment/components/PaymentOptions/Check Check component should render without error 1`] = `
<Check
  changeCheckAmount={[Function]}
  changeRemaining={[Function]}
  clearOptionAndPaymentErrs={[Function]}
  index={0}
  isRefund={false}
  item={
    Object {
      "amount": 0,
      "errors": Array [
        Object {
          "key": 0,
          "name": "checkNumber",
        },
      ],
      "formatCheckAmount": 300,
      "list": Array [
        Object {
          "name": "Credit",
          "value": "172",
        },
        Object {
          "name": "Check",
          "value": "173",
        },
      ],
    }
  }
  onChange={[Function]}
  optionLen={3}
  updateCheckNumber={[Function]}
  value="173"
>
  <div
    className="payment-option payment-check"
  >
    <div
      className="payment-option-list aaui-flexbox"
    >
      <div
        className="aaui-flexbox aaui-flexbox-c1"
      >
        <Dropdown
          autoOpen={false}
          data={
            Array [
              Object {
                "name": "Credit",
                "value": "172",
              },
              Object {
                "name": "Check",
                "value": "173",
              },
            ]
          }
          data-qa-id="payment-paymentOption-optionList"
          filterPlaceholder="Filter..."
          fuzzyQuery={false}
          noResult="No results found."
          onChange={[Function]}
          placeholder="Select one..."
          showAll={true}
          showDeselectall={false}
          showTxtOnlyCheckedOneItem={true}
          theme="flat"
          txtSuffix="selected"
          value="173"
          visible={true}
        >
          <div
            className="aaui-dropdown"
            data-qa-id="payment-paymentOption-optionList"
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
          >
            <button
              className="aaui-dropdown__button aaui-dropdown-btn-flat"
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              role="button"
              type="button"
            >
              <div
                className="aaui-dropdown__button-text"
                title=""
              >
                 Select one...
              </div>
              <i
                className="aaui-dropdown__button-icon icon-chevron-down"
              />
            </button>
            <div
              aria-expanded={false}
              className="aaui-dropdown__menu-wrapper  u-hidden"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="combobox"
              tabIndex={0}
            >
              <div
                className="aaui-dropdown__menu"
              >
                <ul
                  aria-expanded={false}
                  className=""
                  role="listbox"
                >
                  <li
                    className=""
                    key="172"
                    onClick={[Function]}
                    role="option"
                  >
                    <span />
                  </li>
                  <li
                    className=" is-selected"
                    key="173"
                    onClick={[Function]}
                    role="option"
                  >
                    <span />
                  </li>
                </ul>
                <div>
                  <div
                    className="aaui-dropdown__menu-footer"
                  >
                    <div
                      className="u-hidden"
                    >
                      <i
                        className="icon icon-spinner"
                      />
                       Loading...
                    </div>
                    <div />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </Dropdown>
        <Input
          className="payment-check-number"
          data-qa-id="payment-paymentOption-checkNumber"
          errored={true}
          icon={false}
          inputRef={[Function]}
          maxLength={50}
          name="checkNumber"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          size="md"
          type="text"
        >
          <div
            className="input-group input-group--md input-group--error payment-check-number"
          >
            <input
              className="input input-group__field input__field"
              data-qa-id="payment-paymentOption-checkNumber"
              maxLength={50}
              name="checkNumber"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyPress={[Function]}
              onPaste={[Function]}
              type="text"
            />
          </div>
        </Input>
      </div>
      <div
        className="aaui-flexbox aaui-flexbox-c2"
      >
        <label
          className="payment-symbol"
          htmlFor="checkAmount"
        >
          $
        </label>
        <InputNumeric
          allowBlank={true}
          allowKeySpin={true}
          allowMouseSpin={false}
          className=""
          currency="USD"
          data-qa-id="payment-paymentOption-optionAmount"
          decimals={2}
          disabled={false}
          icon=""
          iconHint=""
          id="checkAmount"
          increment={1}
          items={Array []}
          listClass="input-numeric-list"
          listClassName=""
          listConfig={Object {}}
          listMaxHeight="300px"
          listPopupOptions={Object {}}
          listWidth="auto"
          max={9999999999.99}
          min={0}
          name=""
          onBlur={[Function]}
          onFocus={[Function]}
          onListClose={[Function]}
          onListOpen={[Function]}
          onTextChange={[Function]}
          onValueChange={[Function]}
          placeHolder=""
          prefix="an"
          showClear={false}
          showGroup={false}
          showSpinner={false}
          showTrigger={false}
          showTrigger2={false}
          size="md"
          textAlign="right"
          triggerIcon="icon-chevron-down"
          triggerIcon2="icon-chevron-down"
          triggerIconHint=""
          triggerIconHint2=""
          triggerIconToggle="icon-chevron-up"
          triggerIconToggle2="icon-chevron-up"
          value={0}
        >
          <div
            className="an-input-numeric input-group buttons-0 input-group--md"
          >
            <input
              className="input input-group__field textalign-right input__field"
              data-qa-id="payment-paymentOption-optionAmount"
              disabled={false}
              id="checkAmount"
              name=""
              onBlur={[Function]}
              onChange={[Function]}
              onDragEnter={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
              placeholder=""
              type="text"
            />
          </div>
        </InputNumeric>
        <span>
          Check info
        </span>
      </div>
    </div>
    <Alert
      className="payment-amount-exception-alert"
      confirmText="OK"
      onConfirm={[Function]}
      title="System Message"
      type="alert"
    >
      <Modal
        className="modal--alert payment-amount-exception-alert"
        onClose={[Function]}
        shown={false}
        title="System Message"
      >
        <div
          className="modal-wrap"
          role="dialog"
        >
          <div
            className="modal modal--alert payment-amount-exception-alert"
            onKeyDown={[Function]}
          >
            <section
              className="modal-box"
            >
              <header
                className="modal-header"
              >
                <h3
                  className="modal-title"
                >
                  System Message
                </h3>
                <span
                  aria-label="close-dialog"
                  className="icon-close modal-close"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  tabIndex="0"
                />
              </header>
              <div
                className="modal-body"
              />
              <div
                className="modal-footer"
              >
                <Button
                  className=""
                  loading={false}
                  menuData={Array []}
                  noSubmit={false}
                  onClick={[Function]}
                  type="strong"
                >
                  <button
                    className="btn btn-strong"
                    menuData={Array []}
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onMouseOver={[Function]}
                    type="submit"
                  >
                    <span>
                      OK
                    </span>
                  </button>
                </Button>
              </div>
            </section>
          </div>
          <div
            className="modal-mask"
          />
        </div>
      </Modal>
    </Alert>
  </div>
</Check>
`;
